### ISSUE TITLE
Returned anonymous function is not type-checked

### ISSUE THREAD

**Issue by yhslai** [Benjamin-Dobell/IntelliJ-Luanalysis#158]
## Environment

JetBrains Rider 2023.1.3 on Windows 10
 
## Preferences

![image](https://github.com/Benjamin-Dobell/IntelliJ-Luanalysis/assets/538696/d5a721b8-6c50-449c-99f2-67b53a0a102b)

### Lua

Name | Setting  
Lua 5.4

### Type Safety

Strict nil checks | ☑️ 
Unknown type (any) is indexable | ☑️ 
Unknown type (any) is callabale | ☑️ 

## What are the steps to reproduce this issue?

```
---@type fun(): number
return function()
    return "1"
end
```

## What happens?

No warning or error.

## What were you expecting to happen?

Warning or error telling me I'm returning a `fun(): string` instead of `fun(): number`

## Any other comments?

```
---@return fun(): number
return function()
    return "1"
end
```

Doesn't work either.

**Comment by Benjamin-Dobell** [comment]
The `---@return` tag ought to be applied to local or table method function declarations/assignments and pertains to the return type only e.g.

```lua
---@return number
local function test1()
    return "1"
end
```

<img width="193" alt="image" src="https://github.com/Benjamin-Dobell/IntelliJ-Luanalysis/assets/482276/e9f0abb0-eb1f-4b44-81e7-f64b210015fc">

If you're trying to specify a return type, where the return type is a function you'd do:

```lua
---@return fun(): number
local function test2()
    return function()
        return "1"
    end
end
```

<img width="211" alt="image" src="https://github.com/Benjamin-Dobell/IntelliJ-Luanalysis/assets/482276/f3d56bd0-9643-4245-ade6-605e42608267">

Alternatively, if you do want to specify the intended return type inline, you can do so with a `---@type` for the `return` statement e.g.

```lua
local test3 = function()
    ---@type fun(): number
    return function()
        return "1"
    end
end
```

<img width="225" alt="image" src="https://github.com/Benjamin-Dobell/IntelliJ-Luanalysis/assets/482276/b5ffac62-4a8a-43ff-b727-d5e8cf99bdc0">

Basically, the `@return` tag shouldn't appear on `return` statements, rather it's used to specify the return type of the associated function definition.
